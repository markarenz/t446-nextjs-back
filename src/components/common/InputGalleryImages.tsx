import React, { useState, useEffect, useCallback } from 'react';
import {
  DndContext,
  closestCenter,
  MouseSensor,
  TouchSensor,
  DragOverlay,
  useSensor,
  useSensors,
  DragStartEvent,
  DragEndEvent
} from '@dnd-kit/core';
import {
  arrayMove,
  SortableContext,
  rectSortingStrategy
} from '@dnd-kit/sortable';
import Grid from './Grid';
import IconButton from './IconButton';
import ImageUploader from './ImageUploader';
import ImageSelector from './ImageSelector';
import { IconClose, IconNewItem } from '../img/icons';
import SortableItem from './SortableItem';
import Item from './SortableGalleryItem';
import { Asset } from '../../types/types';
import styles from '../../styles/modules/InputGalleryImages.module.scss';

type Props = {
  value: string;
  onChange: Function;
  baseImgUrl: string;
  assetList: Asset[];
  handleTriggerRefresh: Function;
};
const InputGalleryImages: React.FC<Props> = ({
  value,
  onChange,
  baseImgUrl,
  assetList,
  handleTriggerRefresh
}) => {
  const getItemsFromValue = (): string[] => {
    try {
      return JSON.parse(value);
    } catch (err) {
      console.error('items:', err);
      return [];
    }
  };
  const defaultItems = getItemsFromValue();
  const [items, setItems] = useState(defaultItems);
  const [activeId, setActiveId] = useState<string | null>(null);
  const [isAddingImages, setIsAddingImages] = useState(false);
  const sensors = useSensors(useSensor(MouseSensor), useSensor(TouchSensor));

  const handleDragStart = useCallback((event: DragStartEvent) => {
    setActiveId(`${event.active.id}`);
  }, []);
  const handleDragEnd = useCallback((event: DragEndEvent) => {
    const { active, over } = event;

    if (active.id !== over?.id) {
      setItems((items) => {
        const oldIndex = items.indexOf(`${active.id}`);
        const newIndex = items.indexOf(`${over!.id}`);

        return arrayMove(items, oldIndex, newIndex);
      });
    }
    setActiveId(null);
  }, []);
  const handleDragCancel = useCallback(() => {
    setActiveId(null);
  }, []);
  useEffect(() => {
    onChange('images', JSON.stringify(items));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [items]);

  const handleOpenModal = () => {
    document.body.classList.add('modal-open');
    setIsAddingImages(true);
  };
  const handleCloseModal = () => {
    document.body.classList.remove('modal-open');
    setIsAddingImages(false);
  };
  const handleRemoveImg = (filename: string) => {
    console.log('handleRemoveImg', filename);
    setItems((prevItems) => prevItems.filter((i: string) => i !== filename));
  };
  const handleAddItem = (filename: string) => {
    console.log('handleAddImage', filename);
    if (!items.includes(filename)) {
      setItems((prevItems) => [filename, ...items]);
    }
  };

  return (
    <div className={styles.root}>
      <div className={styles.controlWrap}>
        <IconButton
          title="Add Images"
          onClick={handleOpenModal}
          color="secondary"
        >
          <IconNewItem />
        </IconButton>
      </div>

      <DndContext
        sensors={sensors}
        collisionDetection={closestCenter}
        onDragStart={handleDragStart}
        onDragEnd={handleDragEnd}
        onDragCancel={handleDragCancel}
      >
        <SortableContext items={items} strategy={rectSortingStrategy}>
          <Grid columns={5}>
            {items.map((id) => (
              <div className={styles.sortableItemWrap} key={id}>
                <SortableItem id={id} baseImgUrl={baseImgUrl} />
                <div className={styles.btnRemoveWrap}>
                  <IconButton
                    title="Remove"
                    color="primary"
                    onClick={() => handleRemoveImg(id)}
                  >
                    <IconClose />
                  </IconButton>
                </div>
              </div>
            ))}
          </Grid>
        </SortableContext>
        <DragOverlay adjustScale style={{ transformOrigin: '0 0 ' }}>
          {activeId ? (
            <Item id={activeId} isDragging baseImgUrl={baseImgUrl} />
          ) : null}
        </DragOverlay>
      </DndContext>
      {isAddingImages && (
        <div className={styles.modalWrap}>
          <div className={styles.addImages}>
            <IconButton title="close" onClick={handleCloseModal}>
              <IconClose />
            </IconButton>
            <div className={styles.inner}>
              <ImageUploader handleTriggerRefresh={handleTriggerRefresh} />
              <ImageSelector
                allowSelection={true}
                assetList={assetList}
                handleTriggerRefresh={handleTriggerRefresh}
                handleSelection={handleAddItem}
              />
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default InputGalleryImages;
