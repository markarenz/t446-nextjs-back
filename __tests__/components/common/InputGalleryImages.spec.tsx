import React from 'react';
import {
  render,
  screen,
  fireEvent,
  act,
  waitFor
} from '@testing-library/react';
import InputGalleryImages from '../../../src/components/common/InputGalleryImages';
import { getAssetList } from '../../__fixtures__/assetList';

beforeEach(() => {
  jest.resetAllMocks();
});

const defaultProps = {
  value: JSON.stringify(['test1.jpg', 'test2.jpg']),
  onChange: jest.fn(),
  baseImgUrl: 'https://www.domain.com/',
  assetList: getAssetList(),
  handleTriggerRefresh: jest.fn()
};

describe('Button', () => {
  it('renders default button', async () => {
    render(<InputGalleryImages {...defaultProps} />);
    const element = await screen.findByTestId('input-gallery-images');
    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it('handles drag & drop', async () => {
    render(<InputGalleryImages {...defaultProps} />);
    const element = await screen.findByTestId('sortable-0');
    await act(async () => {
      fireEvent.mouseDown(element);
    });
    await act(async () => {
      fireEvent.mouseMove(element, { delta: { x: 200, y: 0 } });
    });
    await act(async () => {
      fireEvent.mouseUp(element);
    });
    expect(defaultProps.onChange).toHaveBeenCalled();
  });

  it('handles removal', async () => {
    render(<InputGalleryImages {...defaultProps} />);
    const element = await screen.findByTestId('btn-remove-gallery-img-0');
    await waitFor(() =>
      fireEvent(
        element,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    await waitFor(
      () =>
        expect(
          screen.getByTestId('sortable-item-test2.jpg')
        ).toBeInTheDocument(),
      {
        timeout: 2000
      }
    );
    const item = await screen.findByTestId('sortable-item-test2.jpg');
    expect(item).toBeInTheDocument();
  });

  it('handles add', async () => {
    render(<InputGalleryImages {...defaultProps} />);
    const btnAdd = await screen.findByTestId('btn-add');
    await act(async () => {
      fireEvent(
        btnAdd,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      );
    });
    await waitFor(
      () => expect(screen.getByTestId('btn-add-close')).toBeInTheDocument(),
      {
        timeout: 2000
      }
    );
    const btnClose = await screen.findByTestId('btn-add-close');
    await act(async () => {
      fireEvent(
        btnClose,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      );
    });
    await act(async () => {
      fireEvent(
        btnAdd,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      );
    });
    const imageItem = await screen.findByTestId('asset-item-0');
    await act(async () => {
      fireEvent(
        imageItem,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      );
    });
    expect(defaultProps.onChange).toHaveBeenCalled();
  });
});
