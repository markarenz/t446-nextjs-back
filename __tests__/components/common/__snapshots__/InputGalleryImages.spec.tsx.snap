// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders default button 1`] = `
<div
  class="root"
  data-testid="input-gallery-images"
>
  <div
    class="controlWrap"
  >
    <button
      class="iconButton secondary"
      data-testid="btn-add"
      title="Add Images"
      type="button"
    >
      <div
        class="inner"
      >
        <svg
          viewBox="0 0 200 200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M100 15a85 85 0 1085 85 84.93 84.93 0 00-85-85zm0 150a65 65 0 1165-65 64.87 64.87 0 01-65 65zm30-72.5h-20v-20a10 10 0 00-20 0v20H70a10 10 0 000 20h20v20a10 10 0 0020 0v-20h20a10 10 0 000-20z"
          />
        </svg>
      </div>
    </button>
  </div>
  <div
    class="root"
    data-testid="grid-wrap"
  >
    <div
      class="sortableItemWrap"
      data-testid="sortable-item-test1.jpg"
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled="false"
        aria-roledescription="sortable"
        class="root "
        data-testid="sortable-0"
        role="button"
        style="background-image: url(https://www.domain.com/thumbnails/test1.jpg);"
        tabindex="0"
      >
        <div
          class="idLabel"
        >
          test1.jpg
        </div>
      </div>
      <div
        class="btnRemoveWrap"
      >
        <button
          class="iconButton primary"
          data-testid="btn-remove-gallery-img-0"
          title="Remove"
          type="button"
        >
          <div
            class="inner"
          >
            <svg
              version="1.1"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M13.182 5L5 13.182 41.818 50 5 86.818 13.182 95 50 58.18 86.818 95 95 86.818 58.18 50 95 13.182 86.818 5 50 41.818z"
                  fill="navy"
                  paint-order="markers fill stroke"
                  stroke="#fff"
                  stroke-linecap="round"
                  stroke-width="0"
                />
              </g>
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      class="sortableItemWrap"
      data-testid="sortable-item-test2.jpg"
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled="false"
        aria-roledescription="sortable"
        class="root "
        data-testid="sortable-1"
        role="button"
        style="background-image: url(https://www.domain.com/thumbnails/test2.jpg);"
        tabindex="0"
      >
        <div
          class="idLabel"
        >
          test2.jpg
        </div>
      </div>
      <div
        class="btnRemoveWrap"
      >
        <button
          class="iconButton primary"
          data-testid="btn-remove-gallery-img-1"
          title="Remove"
          type="button"
        >
          <div
            class="inner"
          >
            <svg
              version="1.1"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M13.182 5L5 13.182 41.818 50 5 86.818 13.182 95 50 58.18 86.818 95 95 86.818 58.18 50 95 13.182 86.818 5 50 41.818z"
                  fill="navy"
                  paint-order="markers fill stroke"
                  stroke="#fff"
                  stroke-linecap="round"
                  stroke-width="0"
                />
              </g>
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</div>
`;
